File: MCPForUnity/Editor/Tools/RenderingStats.cs
using MCPForUnity.Editor.Helpers;
using MCPForUnity.Editor.Tools.Graphics;
using Newtonsoft.Json.Linq;
using Unity.Profiling;
using UnityEditor;
using UnityEngine;
using UnityEngine.Profiling;
using UnityProfiler = UnityEngine.Profiling.Profiler;

namespace MCPForUnity.Editor.Tools
{
    /// <summary>
    /// MCP tool for reading Unity rendering and performance statistics.
    /// Actions: get_stats (rendering counters), get_memory (memory usage),
    ///          get_profiler (profiler frame data).
    /// Most useful during Play mode when rendering is active.
    /// </summary>
    [McpForUnityTool("rendering_stats", AutoRegister = true)]
    public static class RenderingStats
    {
        public static object HandleCommand(JObject @params)
        {
            if (@params == null)
                return new ErrorResponse("Parameters cannot be null.");

            var p = new ToolParams(@params);
            var actionResult = p.GetRequired("action");
            if (!actionResult.IsSuccess)
                return new ErrorResponse(actionResult.ErrorMessage);

            string action = actionResult.Value.ToLowerInvariant();

            try
            {
                return action switch
                {
                    "get_stats"            => GetRenderingStats(p),
                    "get_memory"           => GetMemoryStats(p),
                    "get_profiler"         => GetProfilerStats(p),
                    "get_stats_aggregated" => PerformanceSessionRecorder.GetAggregatedStats(p.GetInt("frames") ?? 0),
                    "get_system_stats"     => PerformanceSessionRecorder.GetSystemStats(p.GetInt("top_n") ?? 10),
                    "get_session_report"   => PerformanceSessionRecorder.GetSessionReport(
                        p.GetBool("include_timeline", true),
                        p.GetBool("include_csv", true)),
                    // Post-session analysis (works after Play mode exit)
                    "list_sessions"        => PerformanceSessionRecorder.ListSessions(),
                    "load_session"         => PerformanceSessionRecorder.LoadSession(p.Get("filename") ?? ""),
                    "analyze_session"      => PerformanceSessionRecorder.AnalyzeSession(p.Get("filename") ?? ""),
                    _ => new ErrorResponse(
                        $"Unknown action: '{action}'. Valid: get_stats, get_memory, get_profiler, " +
                        "get_stats_aggregated, get_system_stats, get_session_report, " +
                        "list_sessions, load_session, analyze_session.")
                };
            }
            catch (System.Exception ex)
            {
                return new ErrorResponse($"Error in {action}: {ex.Message}");
            }
        }

        private static object GetRenderingStats(ToolParams p)
        {
            bool isPlaying = EditorApplication.isPlaying;

            // Compute "saved by batching" = total batched draw calls - total batches
            int savedByBatching =
                (UnityStats.dynamicBatchedDrawCalls - UnityStats.dynamicBatches) +
                (UnityStats.staticBatchedDrawCalls - UnityStats.staticBatches) +
                (UnityStats.instancedBatchedDrawCalls - UnityStats.instancedBatches);

            // CPU/render thread timing via FrameTimingManager — reads previous frame's data,
            // so it works correctly in a single call (no warmup frame needed).
            double cpuMainMs = 0;
            double renderThreadMs = 0;
            FrameTimingManager.CaptureFrameTimings();
            var timings = new FrameTiming[1];
            uint timingCount = FrameTimingManager.GetLatestTimings(1, timings);
            if (timingCount > 0)
            {
                cpuMainMs = timings[0].cpuFrameTime;
                renderThreadMs = timings[0].cpuRenderThreadFrameTime;
            }

            // Animation stats via ProfilerRecorder
            int visibleSkinnedMeshes = 0;
            int animationComponentsPlaying = 0;
            int animatorComponentsPlaying = 0;
            using (var skinnedRec = ProfilerRecorder.StartNew(ProfilerCategory.Render, "Visible Skinned Meshes"))
            {
                if (skinnedRec.Valid) visibleSkinnedMeshes = (int)skinnedRec.LastValue;
            }
            using (var animRec = ProfilerRecorder.StartNew(ProfilerCategory.Animation, "Animation Components Playing"))
            {
                if (animRec.Valid) animationComponentsPlaying = (int)animRec.LastValue;
            }
            using (var animatorRec = ProfilerRecorder.StartNew(ProfilerCategory.Animation, "Animator Components Playing"))
            {
                if (animatorRec.Valid) animatorComponentsPlaying = (int)animatorRec.LastValue;
            }

            var data = new
            {
                isPlaying,
                rendering = new
                {
                    fps = isPlaying ? 1.0f / Time.unscaledDeltaTime : 0f,
                    deltaTime = Time.unscaledDeltaTime,
                    frameCount = Time.frameCount,
                    cpuMainMs,
                    renderThreadMs,
                    drawCalls = UnityStats.drawCalls,
                    batches = UnityStats.batches,
                    savedByBatching,
                    setPassCalls = UnityStats.setPassCalls,
                    triangles = UnityStats.triangles,
                    vertices = UnityStats.vertices,
                    shadowCasters = UnityStats.shadowCasters,
                    visibleSkinnedMeshes,
                    renderTextureChanges = UnityStats.renderTextureChanges,
                    renderTextureBytes = UnityStats.renderTextureBytes,
                    usedTextureMemorySize = UnityStats.usedTextureMemorySize,
                    usedTextureCount = UnityStats.usedTextureCount,
                    screenRes = $"{Screen.width}x{Screen.height}",
                    screenBytes = UnityStats.screenBytes,
                    dynamicBatchedDrawCalls = UnityStats.dynamicBatchedDrawCalls,
                    dynamicBatches = UnityStats.dynamicBatches,
                    staticBatchedDrawCalls = UnityStats.staticBatchedDrawCalls,
                    staticBatches = UnityStats.staticBatches,
                    instancedBatchedDrawCalls = UnityStats.instancedBatchedDrawCalls,
                    instancedBatches = UnityStats.instancedBatches,
                    visible = UnityStats.screenRes != ""
                },
                animation = new
                {
                    animationComponentsPlaying,
                    animatorComponentsPlaying
                },
                audio = new
                {
                    levelDb = AudioListener.volume > 0 ? 20f * Mathf.Log10(AudioListener.volume) : -80f,
                    dspLoad = AudioSettings.dspTime > 0 ? "available" : "unavailable",
                    speakerMode = AudioSettings.speakerMode.ToString(),
                    sampleRate = AudioSettings.outputSampleRate
                },
                qualitySettings = new
                {
                    qualityLevel = QualitySettings.GetQualityLevel(),
                    qualityName = QualitySettings.names[QualitySettings.GetQualityLevel()],
                    vSyncCount = QualitySettings.vSyncCount,
                    targetFrameRate = Application.targetFrameRate,
                    maxQueuedFrames = QualitySettings.maxQueuedFrames
                }
            };

            string note = !isPlaying
                ? "Not in Play mode — most stats will be 0."
                : (data.rendering.visible ? "" : "Game view may not be visible — stats may be 0.");

            return new SuccessResponse(
                isPlaying ? "Rendering stats captured." : "Editor not playing — limited stats.",
                new { data, note }
            );
        }

        private static object GetMemoryStats(ToolParams p)
        {
            const long BytesPerMB = 1024L * 1024L;

            var data = new
            {
                totalAllocatedMemoryMB = UnityProfiler.GetTotalAllocatedMemoryLong() / (double)BytesPerMB,
                totalReservedMemoryMB = UnityProfiler.GetTotalReservedMemoryLong() / (double)BytesPerMB,
                totalUnusedReservedMemoryMB = UnityProfiler.GetTotalUnusedReservedMemoryLong() / (double)BytesPerMB,
                monoUsedSizeMB = UnityProfiler.GetMonoUsedSizeLong() / (double)BytesPerMB,
                monoHeapSizeMB = UnityProfiler.GetMonoHeapSizeLong() / (double)BytesPerMB,
                tempAllocatorSizeMB = UnityProfiler.GetTempAllocatorSize() / (double)BytesPerMB,
                graphicsDriverAllocatedMemoryMB = UnityProfiler.GetAllocatedMemoryForGraphicsDriver() / (double)BytesPerMB,
                isPlaying = EditorApplication.isPlaying
            };

            return new SuccessResponse("Memory stats captured.", data);
        }

        private static object GetProfilerStats(ToolParams p)
        {
            bool isPlaying = EditorApplication.isPlaying;
            int frameCount = Time.frameCount;

            var data = new
            {
                isPlaying,
                frameCount,
                realtimeSinceStartup = Time.realtimeSinceStartup,
                timeSinceLevelLoad = Time.timeSinceLevelLoad,
                timeScale = Time.timeScale,
                fixedDeltaTime = Time.fixedDeltaTime,
                maximumDeltaTime = Time.maximumDeltaTime,
                smoothDeltaTime = Time.smoothDeltaTime,
                captureFramerate = Time.captureFramerate,
                profilerEnabled = UnityProfiler.enabled,
                // System info for context
                systemInfo = new
                {
                    gpuName = SystemInfo.graphicsDeviceName,
                    gpuVendor = SystemInfo.graphicsDeviceVendor,
                    gpuType = SystemInfo.graphicsDeviceType.ToString(),
                    gpuMemoryMB = SystemInfo.graphicsMemorySize,
                    gpuMultiThreaded = SystemInfo.graphicsMultiThreaded,
                    maxTextureSize = SystemInfo.maxTextureSize,
                    renderingThreadingMode = SystemInfo.renderingThreadingMode.ToString(),
                    processorType = SystemInfo.processorType,
                    processorCount = SystemInfo.processorCount,
                    systemMemoryMB = SystemInfo.systemMemorySize
                }
            };

            return new SuccessResponse("Profiler stats captured.", data);
        }
    }
}

File: Server/src/services/tools/rendering_stats.py
"""
Tool for reading Unity rendering statistics, memory usage, and profiler data.
Most useful during Play mode when the rendering pipeline is active.
"""
from typing import Annotated, Any, Literal

from fastmcp import Context
from mcp.types import ToolAnnotations

from services.registry import mcp_for_unity_tool
from services.tools import get_unity_instance_from_context
from transport.unity_transport import send_with_unity_instance
from transport.legacy.unity_connection import async_send_command_with_retry


@mcp_for_unity_tool(
    description=(
        "Read Unity rendering and performance statistics. "
        "Actions: get_stats (single-frame snapshot: draw calls, batches, FPS, cpuMainMs), "
        "get_memory (allocated/reserved/mono/graphics memory), "
        "get_profiler (frame timing, time scale, system info), "
        "get_stats_aggregated (N-frame aggregated: min/max/avg/p50/p95 for FPS, CPU, draw calls — "
        "uses long-lived ProfilerRecorders, much more reliable than single snapshots. "
        "Param: frames=number of recent frames to aggregate, 0=all available), "
        "get_system_stats (per-DOTS-system CPU breakdown sorted by cost — "
        "shows which systems consume the most frame budget. Param: top_n=number of systems), "
        "get_session_report (full Play session report from start to stop — "
        "includes Markdown summary, JSON timeline, CSV. Params: include_timeline=bool, include_csv=bool), "
        "list_sessions (list saved session files from Logs/PerfSessions/ — works anytime), "
        "load_session (load a saved session JSON by filename), "
        "analyze_session (analyze a saved session: bottleneck detection, system ranking, issues. "
        "Param: filename=session JSON file). "
        "Aggregated/system/session actions require Play mode; list/load/analyze work anytime."
    ),
    annotations=ToolAnnotations(
        title="Rendering Stats",
    ),
)
async def rendering_stats(
    ctx: Context,
    action: Annotated[
        Literal[
            "get_stats", "get_memory", "get_profiler",
            "get_stats_aggregated", "get_system_stats", "get_session_report",
            "list_sessions", "load_session", "analyze_session"
        ],
        "Action to perform. get_stats=single snapshot, get_stats_aggregated=N-frame percentiles, "
        "get_system_stats=per-system CPU breakdown, get_session_report=full session timeline+summary, "
        "list_sessions=list saved sessions, load_session=load session JSON, analyze_session=bottleneck analysis."
    ],
    frames: Annotated[int | None, "For get_stats_aggregated: number of recent frames (0=all)."] = None,
    top_n: Annotated[int | None, "For get_system_stats: number of top systems to return."] = None,
    include_timeline: Annotated[bool | None, "For get_session_report: include JSON timeline."] = True,
    include_csv: Annotated[bool | None, "For get_session_report: include CSV data."] = True,
    filename: Annotated[str | None, "For load_session/analyze_session: session JSON filename."] = None,
) -> dict[str, Any]:
    unity_instance = await get_unity_instance_from_context(ctx)

    params: dict[str, Any] = {"action": action}
    if frames is not None:
        params["frames"] = frames
    if top_n is not None:
        params["top_n"] = top_n
    if include_timeline is not None:
        params["include_timeline"] = include_timeline
    if include_csv is not None:
        params["include_csv"] = include_csv
    if filename is not None:
        params["filename"] = filename

    try:
        response = await send_with_unity_instance(
            async_send_command_with_retry, unity_instance,
            "rendering_stats", params
        )

        if isinstance(response, dict) and response.get("success"):
            return {
                "success": True,
                "message": response.get("message", "Stats captured."),
                "data": response.get("data"),
            }
        return response if isinstance(response, dict) else {"success": False, "message": str(response)}

    except Exception as e:
        return {"success": False, "message": f"Python error reading rendering stats: {str(e)}"}
